File: app/src/main/java/ca/mitenko/evn/model/search/Filter.java
package ca.mitenko.evn.model.search;

import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import com.google.android.gms.maps.model.LatLng;

import org.immutables.value.Value;
import org.parceler.Parcel;
import org.parceler.ParcelFactory;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.HashSet;

import ca.mitenko.evn.event.ModifyFilterEvent;
import ca.mitenko.evn.model.Activity;

/**
 * Created by mitenko on 2017-07-30.
 */

@Parcel(value = Parcel.Serialization.VALUE, implementations = ImmutableFilter.class)
@Value.Immutable
public class Filter {

    /**
     * The categories to filter by
     * @return
     */
    @NonNull
    @Value.Default
    public ArrayList<String> categories() {
        return new ArrayList<>();
    }

    /**
     * The activities to filter by
     * @return
     */
    @NonNull
    @Value.Default
    public HashSet<String> activities() {
        return new HashSet<>();
    }

    /**
     * The cost to filter by
     */
    @NonNull
    @Value.Default
    public ArrayList<Integer> cost() {
        return new ArrayList<>();
    }

    /**
     * User Location to sort the list
     */
    @Nullable
    @Value.Default
    public LatLng userLocation() {
        return null;
    }

    /**
     * Returns true / false if the filter is empty
     * @return
     */
    @Value.Lazy
    public boolean isEmpty() {
        return activities().isEmpty() && categories().isEmpty()
                && cost().isEmpty() && userLocation() == null;
    }

    /**
     * Modifies the current filter based on the filter event
     * @param event
     * @return
     */
    public Filter modify(ModifyFilterEvent event, HashMap<String, ArrayList<Activity>> categoryMap) {
        /**
         * Deep copy the current activities and categories
         */
        ArrayList<String> categories = new ArrayList<>();
        categories.addAll(categories());
        HashSet<String> activities = new HashSet<>();
        activities.addAll(activities());
        ArrayList<Integer> costs = new ArrayList<>();
        costs.addAll(cost());

        ModifyFilterEvent.Type type = event.getType();

        if (event.getAction() == ModifyFilterEvent.Action.TOGGLE) {

            if (type == ModifyFilterEvent.Type.CATEGORY) {
                if (categories.contains(event.getValue())) {
                    categories.remove(event.getValue());
                } else {
                    categories.add(event.getValue());
                }

                /**
                 * To properly add activities after a category toggle,
                 * you need to completely redo them
                 */
                activities.clear();
                for(String categoryName : categories) {
                    ArrayList<Activity> catActivities = categoryMap.get(categoryName);
                    for(Activity activity : catActivities) {
                        activities.add(activity.getName());
                    }
                }

            } else if (type == ModifyFilterEvent.Type.ACTIVITY) {
                if (activities.contains(event.getValue())) {
                    activities.remove(event.getValue());
                } else {
                    activities.add(event.getValue());
                }
            } else if (type == ModifyFilterEvent.Type.COST) {
                Integer eventCost = Integer.parseInt(event.getValue());
                if (costs.contains(eventCost)) {
                    costs.remove(eventCost);
                } else {
                    costs.add(eventCost);
                }
            }
        }

        if (event.getAction() == ModifyFilterEvent.Action.CLEAR_ALL) {
            categories.clear();
            activities.clear();
        }

        return ImmutableFilter.builder()
                .categories(categories)
                .activities(activities)
                .cost(costs)
                .build();
    }

    /**
     * Parcel Factory
     * @return
     */
    @ParcelFactory
    public static Filter build(HashSet<String> activities, ArrayList<Integer> cost,
        ArrayList<String> categories, LatLng userLocation) {
        return ImmutableFilter.builder()
                .activities(activities)
                .categories(categories)
                .userLocation(userLocation)
                .cost(cost)
                .build();
    }
}

File: app/src/main/java/ca/mitenko/evn/model/search/Filter.kt
package ca.mitenko.evn.model.search

import android.os.Parcelable
import ca.mitenko.evn.event.ModifyFilterEvent
import ca.mitenko.evn.model.Activity
import com.google.android.gms.maps.model.LatLng
import kotlinx.android.parcel.Parcelize
import org.immutables.value.Value
import java.util.*
import kotlin.collections.ArrayList

/**
 * Created by mitenko on 2017-07-30.
 */

@Parcelize
data class Filter(
        val categories: ArrayList<String> = ArrayList(),
        val activities: HashSet<String> = HashSet(),
        val cost: ArrayList<Int> = ArrayList(),
        val userLocation: LatLng? = null
): Parcelable {

    /**
     * Returns true / false if the filter is empty
     * @return
     */
    open val isEmpty: Boolean
        @Value.Lazy
        get() = (activities.isEmpty() && categories.isEmpty()
                && cost.isEmpty() && userLocation == null)
    /**
     * Modifies the current filter based on the filter event
     * @param event
     * @return
     */
    fun modify(event: ModifyFilterEvent, categoryMap: HashMap<String, ArrayList<Activity>>): Filter {
        /**
         * Deep copy the current activities and categories
         */
        val categories = ArrayList<String>()
        categories.addAll(categories)
        val activities = HashSet<String>()
        activities.addAll(activities)
        val costs = ArrayList<Int>()
        costs.addAll(cost)

        val type = event.type

        if (event.action === ModifyFilterEvent.Action.TOGGLE) {

            if (type === ModifyFilterEvent.Type.CATEGORY) {
                if (categories.contains(event.value)) {
                    categories.remove(event.value)
                } else {
                    categories.add(event.value)
                }

                /**
                 * To properly add activities after a category toggle,
                 * you need to completely redo them
                 */
                activities.clear()
                for (categoryName in categories) {
                    val catActivities = categoryMap[categoryName]
                    if (catActivities != null && !catActivities.isEmpty()) {
                        for ((_, name) in catActivities) {
                            activities.add(name)
                        }
                    }
                }

            } else if (type === ModifyFilterEvent.Type.ACTIVITY) {
                if (activities.contains(event.value)) {
                    activities.remove(event.value)
                } else {
                    activities.add(event.value)
                }
            } else if (type === ModifyFilterEvent.Type.COST) {
                val eventCost = Integer.parseInt(event.value)
                if (costs.contains(eventCost)) {
                    costs.remove(eventCost)
                } else {
                    costs.add(eventCost)
                }
            }
        }

        if (event.action === ModifyFilterEvent.Action.CLEAR_ALL) {
            categories.clear()
            activities.clear()
        }

        return this.copy(categories = categories, activities = activities, cost = costs)
    }
}
